src/sleep/sleep.c

```diff
@@ -263,41 +263,107 @@ static int execute(
+                if (r >= 0) {
+                        current_capacity = r;
+                        log_debug("Current battery charge percentage after wakeup: %d%%", current_capacity);
+                } else if (r == -ENOENT) {
+                        log_debug_errno(r, "Battery capacity percentage unavailable, cannot estimate discharge rate: %m");
+                        break;
+                        /* In case of no battery, system will be hibernated after 1st cycle of suspend */
+                } else
+                        return log_error_errno(r, "Error fetching battery capacity percentage: %m");
+
+                if (!woken_by_timer) {
+                        if (current_capacity < last_capacity) {
+                                /* We woke up before alarm time, estimate discharge rate using suspension duration. */
+                                after_timestamp = now(CLOCK_BOOTTIME);
+                                log_debug("Attempting to estimate battery discharge rate after manual wakeup");
+                                estimated_discharge_rate = (last_capacity - current_capacity) * USEC_PER_HOUR / (after_timestamp - before_timestamp);
+                                r = put_battery_discharge_rate(estimated_discharge_rate);
+                                if (r < 0)
+                                        log_error_errno(r, "Failed to update battery discharge rate: ignoring %m");
+                        }
+                        return 0; /* return as manual wakeup done. This also will return in case battery was charged during suspension */
+                }
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                if (current_capacity >= last_capacity)
+                        log_debug("Battery was not discharged during suspension");
+                else {
+                        log_debug("Attempting to estimate battery discharge rate after auto wakeup from %s hours sleep timer",
+                                  FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));
 
-        tfd = safe_close(tfd);
+                        /* If woken up after alarm time, estimate discharge rate for suspend interval */
+                        estimated_discharge_rate = (last_capacity - current_capacity) * USEC_PER_HOUR / suspend_interval;
 
-        /* If woken up after alarm time, hibernate */
-        log_debug("Attempting to hibernate after waking from %s timer",
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+                        r = put_battery_discharge_rate(estimated_discharge_rate);
+                        if (r < 0)
+                                log_error_errno(r, "Failed to update battery discharge rate, ignoring: %m");
+                }
```

Comment:
Even if triggered by the timer, please use the current time. How about the following?
```c
if (current_capacity >= last_capacity)
        log_debug("Battery was not discharged during suspension.");
else {
        /* We woke up before alarm time, estimate discharge rate using suspension duration. */
        after_timestamp = now(CLOCK_BOOTTIME);
        estimated_discharge_rate = (last_capacity - current_capacity) * USEC_PER_HOUR / (after_timestamp - before_timestamp);
        r = put_battery_discharge_rate(estimated_discharge_rate);
        if (r < 0)
                log_error_warning(r, "Failed to save battery discharge rate: ignoring %m");
}

if (!woken_by_timer)
        return 0;
```

Reply:
It makes sense to avoid any sort of discrepancy and use current time instead of duration of suspension but then there is no benefit in checking we woke automatically or manually because eventually we are doing same thing then in both cases. Don't you think ?

Reply:
Maybe. But not sure. I have not followed why you introduced the flag.

Reply:
we need the flag to handle manual wakeup. but rest of it could use the current time to get the discharge rate. It makes sense. Thanks for highlighting. i will fix.